pkg/kv/kvclient/kvcoord/dist_sender.go

```diff
@@ -2557,71 +2543,83 @@ func (ds *DistSender) sendToReplicas(
 	if regionErr != nil {
-		log.VEventf(ctx, 2, "%v", regionErr)
+		log.VEventf(ctx, 2, "unable to determine if batch is cross region %v", regionErr)
 	}
 	if zoneErr != nil {
-		log.VEventf(ctx, 2, "%v", zoneErr)
+		log.VEventf(ctx, 2, "unable to determine if batch is cross zone %v", zoneErr)
 	}
-	return isCrossRegion, isCrossZone
+	return comparisonResult
 }
 
-// checkAndUpdateBatchRequestMetrics updates the batch requests metrics in a
-// more meaningful way. Cross-region metrics monitor activities across different
-// regions. Cross-zone metrics monitor cross-zone activities within the same
-// region or in cases where region tiers are not configured. The check result is
-// returned here to avoid redundant check for metrics updates after receiving
-// batch responses.
-func (ds *DistSender) checkAndUpdateBatchRequestMetrics(
+// checkAndUpdateCrossLocalityBatchMetrics updates the batch requests metrics in
+// a more meaningful way. Cross-region metrics monitor activities across
+// different regions. Cross-zone metrics monitor cross-zone activities within
+// the same region or in cases where region tiers are not configured. The
+// locality comparison result is returned here to avoid redundant check for
+// metrics updates after receiving batch responses.
+func (ds *DistSender) checkAndUpdateCrossLocalityBatchMetrics(
 	ctx context.Context, ba *kvpb.BatchRequest,
-) (shouldIncCrossRegion bool, shouldIncCrossZone bool) {
-	ds.metrics.ReplicaAddressedBatchRequestBytes.Inc(int64(ba.Size()))
-	isCrossRegion, isCrossZone := ds.isCrossRegionCrossZoneBatch(ctx, ba)
-	if isCrossRegion {
-		if !isCrossZone {
-			log.VEventf(ctx, 2, "unexpected: cross region but same zone")
-		} else {
-			ds.metrics.CrossRegionBatchRequestBytes.Inc(int64(ba.Size()))
-			shouldIncCrossRegion = true
-		}
-	} else {
-		if isCrossZone {
-			ds.metrics.CrossZoneBatchRequestBytes.Inc(int64(ba.Size()))
-			shouldIncCrossZone = true
-		}
+) roachpb.LocalityComparisonType {
+	if ds.TestingBatchInterceptor != nil {
+		// TestingBatchInterceptor intercepts DistSender.Send() with the purpose of
+		// passing the actual byte count of batch size to tests to verify the
+		// accuracy of updates made to DistSender metrics.
+		ds.TestingBatchInterceptor(ba, nil)
 	}
-	return shouldIncCrossRegion, shouldIncCrossZone
+
+	ds.metrics.ReplicaAddressedBatchRequestBytes.Inc(int64(ba.Size()))
+	comparisonResult := ds.getCrossLocalityComparison(ctx, ba)
+	switch comparisonResult {
+	case roachpb.LocalityComparisonType_CROSS_REGION_CROSS_ZONE:
+		ds.metrics.CrossRegionBatchRequestBytes.Inc(int64(ba.Size()))
+	case roachpb.LocalityComparisonType_SAME_REGION_CROSS_ZONE:
+		ds.metrics.CrossZoneBatchRequestBytes.Inc(int64(ba.Size()))
+	case roachpb.LocalityComparisonType_CROSS_REGION_SAME_ZONE:
+		log.VEventf(ctx, 2, "unexpected: cross region but same zone")
+	case roachpb.LocalityComparisonType_SAME_REGION_SAME_ZONE:
+		// No metrics or error reporting.
+	}
```

Comment:
The default case here was intended to handle cases where the given locality comparison type `roachpb.LocalityComparisonType_UNDEFINED` which could happen if `getLocalityComparison` fails to fetch node descriptors. I feel this could happen in test cases where nothing is configured properly. I could add another switch case to just return here. Adding a panic on a metrics update helper function seems to be too disruptive.